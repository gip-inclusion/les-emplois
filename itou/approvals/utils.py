from itou.companies.enums import SIAE_WITH_CONVENTION_KINDS


def get_user_last_accepted_siae_job_application(user):
    if not user.is_job_seeker:
        return None

    # Some candidates may not have accepted job applications
    # Assuming it's the case can lead to issues downstream
    return (
        user.job_applications.accepted()
        .filter(to_company__kind__in=SIAE_WITH_CONVENTION_KINDS)
        .with_accepted_at()
        .order_by("-accepted_at", "-hiring_start_at")
        .first()
    )


def last_hire_was_made_by_siae(user, siae):
    if not user.is_job_seeker:
        return False
    last_accepted_job_application = get_user_last_accepted_siae_job_application(user)
    return last_accepted_job_application and last_accepted_job_application.to_company_id == siae.pk
