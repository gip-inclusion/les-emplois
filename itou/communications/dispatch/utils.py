class JobSeekerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_job_seeker


class EmployerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_employer


class PrescriberNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_prescriber


class PrescriberOrEmployerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and (self.user.is_prescriber or self.user.is_employer)
