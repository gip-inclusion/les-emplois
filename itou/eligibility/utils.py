from collections import Counter


def geiq_allowance_amount(is_authorized_prescriber, administrative_criteria) -> int:
    """Amount of granted allowance for job seeker.

    Calculated in function of:
        - author kind
        - number, annex and level of administrative criteria.
    Currently, only 3 amounts possible:
        - 0
        - 814EUR
        - 1400EUR
    """

    # Even when no criteria in the diagnosis
    if is_authorized_prescriber:
        return 1400

    # Count by annex
    annex_cnt = Counter(c.annex for c in administrative_criteria)

    # Only annex 2 administrative criteria have a level defined
    level_cnt = Counter(c.level for c in administrative_criteria if c.level)

    # At least one level 1 criterion or at least two level 2 criteria
    if level_cnt["1"] > 0 or level_cnt["2"] > 1:
        return 1400

    # Criteria in both annex ("1+2") must be counted as annex 1
    if annex_cnt["1"] + annex_cnt["1+2"] > 0:
        return 814

    return 0
