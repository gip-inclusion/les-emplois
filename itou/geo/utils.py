from typing import Optional

from django.contrib.gis.geos import GEOSGeometry


def coords_to_geometry(lat, lon) -> Optional[GEOSGeometry]:
    if lat is not None and lon is not None:
        return GEOSGeometry(f"POINT({lon} {lat})")
    return None


def multipolygon_to_geometry(spec: str) -> GEOSGeometry:
    return GEOSGeometry(f"{spec}")
