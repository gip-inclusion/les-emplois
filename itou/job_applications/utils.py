def show_afpa_ad(user):
    postcode = user.jobseeker_profile.hexa_post_code or user.post_code
    return postcode[:2] in [
        # Hauts-de-france
        "02",
        "59",
        "60",
        "62",
        "80",
        # Nouvelle aquitaine
        "16",
        "17",
        "19",
        "23",
        "24",
        "33",
        "40",
        "47",
        "64",
        "79",
        "86",
        "87",
    ]
