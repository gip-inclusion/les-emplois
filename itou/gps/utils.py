import logging
from typing import NamedTuple

import pandas as pd
from django.core.cache import cache

from itou.gps.models import FranceTravailContact
from itou.utils.python import timeit


logger = logging.getLogger(__name__)


class FranceTravailContactDetails(NamedTuple):
    prescriber_name: str
    prescriber_email: str


GPS_ADVISORS_KEY = "GPS_ADVISORS"


@timeit
def parse_gps_advisors_file(import_file):
    # NOM | PRENOM | NIR | DATE DE NAISSANCE | ASSEDIC | DC_STRUCTUREPRINCIPALEDE | DC_AGENTREFERENT
    # DC_STRUCTURERATTACH | DC_NOMAGENTREFERENT | DC_MAIL | DC_LBLPOSITIONNEMENTIAE
    df = pd.read_excel(
        import_file,
        converters={
            "NIR": str,
            "DC_NOMAGENTREFERENT": str,
            "DC_MAIL": str,
        },
    )

    df = df.rename(columns={"NIR": "nir", "DC_NOMAGENTREFERENT": "prescriber_name", "DC_MAIL": "prescriber_email"})
    df = df[["nir", "prescriber_name", "prescriber_email"]]
    df = df.dropna(subset=["prescriber_name", "prescriber_email"])
    logger.info(f"Found {len(df)} rows from GPS export.")

    nir_to_contact = {}
    count_nir_invalid_length = 0
    for row in df.itertuples():
        nir = row.nir
        match len(nir):
            case 13:
                # not all of the NIRs in the dataset are complete, so we treat them
                nir = f"{nir}{str(97 - int(nir) % 97).zfill(2)}"
            case 15:
                pass
            case _:
                count_nir_invalid_length += 1
                continue
        contact_details = FranceTravailContactDetails(row.prescriber_name, row.prescriber_email)
        existing = nir_to_contact.setdefault(nir, contact_details)
        if existing != contact_details:
            logger.warning(
                f"JobSeekerProfile with nir {nir} matched an additional contact that have not been imported."
            )

    if count_nir_invalid_length:
        logger.warning(f"There are {count_nir_invalid_length} included NIR values of invalid length after treatment.")

    cache.set(GPS_ADVISORS_KEY, nir_to_contact)
    return nir_to_contact


def create_or_update_advisor(jobseeker_profile, nir_to_contact, commit=True):
    if jobseeker_profile.nir not in nir_to_contact:
        return None, False

    contact_name, contact_email = nir_to_contact[jobseeker_profile.nir]
    created = False
    # prepare to create or update FranceTravailContact
    try:
        advisor = jobseeker_profile.advisor_information
        advisor.name = contact_name
        advisor.email = contact_email
    except FranceTravailContact.DoesNotExist:
        advisor = FranceTravailContact(
            name=contact_name,
            email=contact_email,
            jobseeker_profile=jobseeker_profile,
        )
        created = True
    if commit:
        advisor.save()

    return advisor, created


def find_job_seeker_advisor(jobseeker_profile):
    if not jobseeker_profile.nir:
        return

    nir_to_contact = cache.get(GPS_ADVISORS_KEY)

    # No cache available : wait for the next cron call (in less than 15 minutes)
    if nir_to_contact is None:
        return

    create_or_update_advisor(jobseeker_profile, nir_to_contact)
