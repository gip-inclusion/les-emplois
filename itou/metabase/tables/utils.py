import functools
import hashlib
from operator import attrgetter

from django.conf import settings
from django.db.models import JSONField
from django.db.models.fields import (
    AutoField,
    BooleanField,
    CharField,
    DateField,
    DateTimeField,
    PositiveIntegerField,
    TextField,
    UUIDField,
)
from django.db.models.fields.related import ForeignKey
from django.utils import timezone

from itou.approvals.enums import Origin, ProlongationReason
from itou.approvals.models import Approval
from itou.cities.models import City
from itou.common_apps.address.departments import DEPARTMENT_TO_REGION, DEPARTMENTS
from itou.common_apps.address.models import BAN_API_RELIANCE_SCORE
from itou.companies.models import Company
from itou.geo.enums import ZRRStatus
from itou.geo.models import ZRR
from itou.job_applications.models import JobApplicationWorkflow
from itou.users.models import User


class MetabaseTable:
    def __init__(self, name):
        self.name = name
        self.columns = []

    def add_columns(self, columns):
        self.columns += columns

    def get(self, column_name, input):
        matching_columns = [c for c in self.columns if c["name"] == column_name]
        assert len(matching_columns) == 1
        fn = matching_columns[0]["fn"]
        return fn(input)


def get_field_type_from_field(field):
    if isinstance(field, CharField):
        return "varchar"
    if isinstance(field, TextField):
        return "text"
    if isinstance(field, PositiveIntegerField):
        return "integer"
    if isinstance(field, AutoField) and field.name == "id":
        return "integer"
    if isinstance(field, UUIDField):
        return "uuid"
    if isinstance(field, DateTimeField):
        return "timestamp with time zone"
    if isinstance(field, DateField):
        return "date"
    if isinstance(field, ForeignKey):
        related_pk_field = field.related_model._meta.pk
        return get_field_type_from_field(related_pk_field)
    if isinstance(field, JSONField):
        return "jsonb"
    if isinstance(field, BooleanField):
        return "boolean"
    raise ValueError("Unexpected field type")


def get_column_from_field(field, name):
    """
    Guess column configuration for simple fields with no subtlety.
    """
    field_name = field.name
    if isinstance(field, ForeignKey):
        field_name += "_id"
    return {
        "name": name,
        "type": get_field_type_from_field(field),
        "comment": field.verbose_name,
        "fn": lambda o: getattr(o, field_name),
    }


def get_choice(choices, key):
    choices = dict(choices)
    # Gettext fixes `can't adapt type '__proxy__'` error
    # due to laxy_gettext and psycopg not going well together.
    # See https://code.djangoproject.com/ticket/13965
    if key in choices:
        # FIXME: we dropped translations, this should be easier now.
        # return gettext(choices[key])
        return choices[key]
    return None


def get_first_membership_join_date(memberships):
    memberships = list(memberships.all())
    # We have to do all this in python to benefit from prefetch_related.
    if len(memberships) >= 1:
        memberships.sort(key=lambda o: o.joined_at)
        first_membership = memberships[0]
        return first_membership.joined_at
    return None


def get_hiring_company(job_seeker):
    """
    Ideally the job_seeker would have a unique hiring so that we can
    properly link the approval back to the company. However we already
    have many job_seekers with two or more hirings. In this case
    we consider the latest hiring, which is an ugly workaround
    around the fact that we do not have a proper approval=>siae
    link yet.
    """
    assert job_seeker.is_job_seeker
    hirings = [ja for ja in job_seeker.job_applications.all() if ja.state == JobApplicationWorkflow.STATE_ACCEPTED]
    if hirings:
        latest_hiring = max(hirings, key=attrgetter("created_at"))
        return latest_hiring.to_company
    return None


def get_department_and_region_columns(name_suffix="", comment_suffix="", custom_fn=lambda o: o):
    return [
        {
            "name": f"département{name_suffix}",
            "type": "varchar",
            "comment": f"Département{comment_suffix}",
            "fn": lambda o: getattr(custom_fn(o), "department", None),
        },
        {
            "name": f"nom_département{name_suffix}",
            "type": "varchar",
            "comment": f"Nom complet du département{comment_suffix}",
            "fn": lambda o: DEPARTMENTS.get(getattr(custom_fn(o), "department", None)),
        },
        {
            "name": f"région{name_suffix}",
            "type": "varchar",
            "comment": f"Région{comment_suffix}",
            "fn": lambda o: DEPARTMENT_TO_REGION.get(getattr(custom_fn(o), "department", None)),
        },
    ]


# FIXME @dejafait drop this as soon as data analysts no longer use
# structures_v0.code_commune nor organisations.code_commune
# because one post_code can actually have *several* insee_codes (╯°□°)╯︵ ┻━┻
@functools.cache
def get_post_code_to_insee_code_map():
    """
    Load once and for all this ~35k items dataset in memory.
    """
    post_code_to_insee_code_map = {}
    for city in City.objects.all():
        for post_code in city.post_codes:
            post_code_to_insee_code_map[post_code] = city.code_insee
    return post_code_to_insee_code_map


# FIXME @dejafait drop this as soon as data analysts no longer use
# structures_v0.code_commune nor organisations.code_commune
# because one post_code can actually have *several* insee_codes (╯°□°)╯︵ ┻━┻
def convert_post_code_to_insee_code(post_code):
    return get_post_code_to_insee_code_map().get(post_code)


@functools.cache
def get_insee_code_to_zrr_status_map():
    """
    Load once and for all this ~35k items dataset in memory.
    """
    insee_code_to_zrr_status_map = {}
    for zrr in ZRR.objects.all():
        insee_code_to_zrr_status_map[zrr.insee_code] = zrr.status
    return insee_code_to_zrr_status_map


def get_zrr_status_for_insee_code(insee_code):
    raw_zrr_status = get_insee_code_to_zrr_status_map().get(insee_code)
    if raw_zrr_status:
        return ZRRStatus(raw_zrr_status).label
    return "Statut ZRR inconnu"


def get_post_code_column(name_suffix="", comment_suffix="", custom_fn=lambda o: o):
    return {
        "name": f"code_postal{name_suffix}",
        "type": "varchar",
        "comment": f"Code postal{comment_suffix}",
        "fn": lambda o: custom_fn(o).post_code,
    }


def get_address_columns(name_suffix="", comment_suffix="", custom_fn=lambda o: o):
    return (
        [
            {
                "name": f"adresse_ligne_1{name_suffix}",
                "type": "varchar",
                "comment": f"Première ligne adresse{comment_suffix}",
                "fn": lambda o: custom_fn(o).address_line_1,
            },
            {
                "name": f"adresse_ligne_2{name_suffix}",
                "type": "varchar",
                "comment": f"Seconde ligne adresse{comment_suffix}",
                "fn": lambda o: custom_fn(o).address_line_2,
            },
        ]
        + [get_post_code_column(name_suffix, comment_suffix, custom_fn)]
        + [
            {
                # FIXME @dejafait drop this as soon as data analysts no longer use
                # structures_v0.code_commune nor organisations.code_commune
                # because one post_code can actually have *several* insee_codes (╯°□°)╯︵ ┻━┻
                "name": f"code_commune{name_suffix}",
                "type": "varchar",
                "comment": f"Code commune{comment_suffix}",
                "fn": lambda o: convert_post_code_to_insee_code(custom_fn(o).post_code),
            },
            {
                "name": f"ville{name_suffix}",
                "type": "varchar",
                "comment": f"Ville{comment_suffix}",
                "fn": lambda o: custom_fn(o).city,
            },
            {
                "name": f"longitude{name_suffix}",
                "type": "double precision",
                "comment": f"Longitude{comment_suffix}",
                "fn": lambda o: custom_fn(o).longitude,
            },
            {
                "name": f"latitude{name_suffix}",
                "type": "double precision",
                "comment": f"Latitude{comment_suffix}",
                "fn": lambda o: custom_fn(o).latitude,
            },
        ]
        + get_department_and_region_columns(name_suffix, comment_suffix, custom_fn)
    )


def get_establishment_last_login_date_column():
    return [
        {
            "name": "date_dernière_connexion",
            "type": "date",
            "comment": "Date de dernière connexion utilisateur",
            "fn": lambda o: max([u.last_login for u in o.members.all() if u.last_login], default=None)
            if o.members.exists()
            else None,
        },
    ]


def get_establishment_is_active_column():
    return [
        {
            "name": "active",
            "type": "boolean",
            "comment": "Dernière connexion dans les 7 jours",
            "fn": lambda o: any(
                [u.last_login > timezone.now() - timezone.timedelta(days=7) for u in o.members.all() if u.last_login]
            )
            if o.members.exists()
            else False,
        },
    ]


@functools.cache
def get_ai_stock_job_seeker_pks():
    return Approval.objects.filter(origin=Origin.AI_STOCK).values_list("user_id", flat=True).distinct()


@functools.cache
def get_active_companies_pks():
    """
    Load once and for all the list of all active company pks in memory and reuse them multiple times in various
    queries to avoid additional joins of the SiaeConvention model and the non trivial use of the
    `Company.objects.active()` queryset on a related model of a queryset on another model. This is a list of less
    than 10k integers thus should not use much memory. The end result being both simpler code
    and better performance.
    """
    return list(Company.objects.active().values_list("pk", flat=True))


@functools.cache
def get_qpv_job_seeker_pks():
    """
    Load once and for all the list of all job seeker pks which are located in a QPV zone.

    The alternative would have been to naively compute `QPV.in_qpv(u, geom_field="coords")` for each and every one
    of the ~700k job seekers, which would have resulted in a undesirable deluge of 700k micro SQL requests.

    Unfortunately we failed so far at finding a clean ORM friendly way to do this in a single SQL request.
    """
    qpv_job_seekers = User.objects.raw(
        # Takes only ~2s on local dev.
        "SELECT uu.id FROM users_user uu "
        "INNER JOIN geo_qpv gq ON ST_Contains(gq.geometry, uu.coords::geometry) "
        f"WHERE uu.coords IS NOT NULL AND uu.geocoding_score > {BAN_API_RELIANCE_SCORE}"
    )
    # A list of ~100k integers is permanently loaded in memory. It is fortunately not a very high volume of data.
    # Objects returned by `raw` are defered which means their fields are not preloaded unless they have been
    # explicitely specified in the SQL request. We did specify and thus preload `id` fields.
    return [job_seeker.pk for job_seeker in qpv_job_seekers]


def hash_content(content):
    return hashlib.sha256(f"{content}{settings.METABASE_HASH_SALT}".encode()).hexdigest()


def get_common_prolongation_columns(get_field_fn):
    return [
        get_column_from_field(get_field_fn("id"), name="id"),
        get_column_from_field(get_field_fn("approval"), name="id_pass_agrément"),
        get_column_from_field(get_field_fn("start_at"), name="date_début"),
        get_column_from_field(get_field_fn("end_at"), name="date_fin"),
        {
            "name": "motif",
            "type": "varchar",
            "comment": "Motif renseigné",
            "fn": lambda o: get_choice(choices=ProlongationReason.choices, key=o.reason),
        },
        # Do not inject `reason_explanation` as it contains highly sensitive personal information in practice.
        get_column_from_field(get_field_fn("declared_by"), name="id_utilisateur_déclarant"),
        get_column_from_field(get_field_fn("declared_by_siae"), name="id_structure_déclarante"),
        get_column_from_field(get_field_fn("validated_by"), name="id_utilisateur_prescripteur"),
        get_column_from_field(get_field_fn("prescriber_organization"), name="id_organisation_prescripteur"),
    ]
