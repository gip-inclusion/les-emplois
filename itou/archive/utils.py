import datetime

from django.db import models
from django.db.models import Count, Exists, OuterRef, Q
from django.db.models.functions import Coalesce
from django.utils import timezone

from itou.approvals.models import Approval
from itou.eligibility.models import EligibilityDiagnosis, GEIQEligibilityDiagnosis
from itou.gps.models import FollowUpGroup
from itou.job_applications.models import JobApplication
from itou.users.models import User, UserKind


def get_filter_kwargs_on_user_for_related_objects_to_check():
    """
    Returns a dictionary of filter parameters to check for objects related
    to the User model that are not cascade deleted.
    """
    return {
        f"{obj.name}__isnull": True
        for obj in User._meta.related_objects
        if getattr(obj, "on_delete", None) and obj.on_delete != models.CASCADE
    }


def get_year_month_or_none(date=None):
    if not date:
        return None

    if isinstance(date, datetime.datetime):
        return timezone.localdate(date).replace(day=1)

    return date.replace(day=1)


def count_related_subquery(model, fk_field, outer_ref_field, extra_filters=None):
    filters = {fk_field: OuterRef(outer_ref_field)}
    if extra_filters:
        filters.update(extra_filters)
    return Coalesce(
        model.objects.filter(**filters).values(fk_field).annotate(count=Count(outer_ref_field)).values("count"), 0
    )


def inactive_jobseekers_without_recent_related_objects(inactive_since, notified, batch_size=None):
    recent_approval = Approval.objects.filter(
        Q(end_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), user_id=OuterRef("pk")
    )
    recent_eligibility_diagnosis = EligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_geiq_eligibility_diagnosis = GEIQEligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_job_application = JobApplication.objects.filter(job_seeker_id=OuterRef("pk"), updated_at__gt=inactive_since)
    recent_followup_group = FollowUpGroup.objects.filter(beneficiary_id=OuterRef("pk"), updated_at__gt=inactive_since)

    qs = (
        User.objects.filter(
            (Q(last_login__lte=inactive_since) | Q(last_login__isnull=True)),
            date_joined__lte=inactive_since,
            kind=UserKind.JOB_SEEKER,
            upcoming_deletion_notified_at__isnull=not notified,
        )
        .filter(
            ~Exists(recent_approval),
            ~Exists(recent_eligibility_diagnosis),
            ~Exists(recent_geiq_eligibility_diagnosis),
            ~Exists(recent_job_application),
            ~Exists(recent_followup_group),
        )
        .order_by("date_joined", "pk")
    )

    if batch_size:
        return qs[:batch_size]

    return qs
