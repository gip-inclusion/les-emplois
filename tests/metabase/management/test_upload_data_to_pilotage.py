import io

import pytest
from django.core.management import call_command
from django.utils import timezone

from itou.utils.storage.s3 import pilotage_s3_client


@pytest.mark.usefixtures("temporary_bucket")
def test_command_call(tmp_path, snapshot, caplog):
    tmp_path.joinpath("fluxIAE_ITOU_.tar.gz").write_bytes(b"Hello World")
    call_command("upload_data_to_pilotage", tmp_path.as_posix(), "--wet-run")
    assert caplog.messages[:-1] == snapshot
    assert caplog.messages[-1].startswith(
        "Management command itou.metabase.management.commands.upload_data_to_pilotage succeeded in"
    )


@pytest.mark.usefixtures("temporary_bucket")
def test_command_idempotence(tmp_path, settings):
    filename = f"fluxIAE_ITOU_{timezone.now():%Y%m%d_%H%M%S}.tar.gz"
    tmp_path.joinpath(filename).touch()

    call_command("upload_data_to_pilotage", tmp_path.as_posix(), "--wet-run")
    response = pilotage_s3_client().list_objects_v2(
        Bucket=settings.PILOTAGE_DATASTORE_S3_BUCKET_NAME,
    )
    assert response["KeyCount"] == 1
    assert response["Contents"][0]["Key"] == f"flux-iae/{filename}"
    modified_at = response["Contents"][0]["LastModified"]

    call_command("upload_data_to_pilotage", tmp_path.as_posix(), "--wet-run")
    response = pilotage_s3_client().list_objects_v2(
        Bucket=settings.PILOTAGE_DATASTORE_S3_BUCKET_NAME,
    )
    assert response["KeyCount"] == 1
    assert response["Contents"][0]["Key"] == f"flux-iae/{filename}"
    assert response["Contents"][0]["LastModified"] == modified_at


@pytest.mark.usefixtures("temporary_bucket")
def test_command_check_content_length(caplog, snapshot, tmp_path, settings):
    client = pilotage_s3_client()
    filename = "fluxIAE_ITOU_.tar.gz"
    tmp_path.joinpath(filename).write_bytes(b"Hello World")

    client.upload_fileobj(io.BytesIO(), Bucket=settings.PILOTAGE_DATASTORE_S3_BUCKET_NAME, Key=f"flux-iae/{filename}")
    call_command("upload_data_to_pilotage", tmp_path.as_posix(), "--wet-run")
    assert (
        client.get_object(Bucket=settings.PILOTAGE_DATASTORE_S3_BUCKET_NAME, Key=f"flux-iae/{filename}")["Body"].read()
        == b"Hello World"
    )
    assert caplog.messages[:-1] == snapshot
    assert caplog.messages[-1].startswith(
        "Management command itou.metabase.management.commands.upload_data_to_pilotage succeeded in"
    )


def test_command_retries(caplog, snapshot, tmp_path, settings, mocker):
    filename = "fluxIAE_ITOU_.tar.gz"
    tmp_path.joinpath(filename).touch()
    mocker.patch("itou.metabase.management.commands.upload_data_to_pilotage.Command._upload_file", return_value=None)

    call_command("upload_data_to_pilotage", tmp_path.as_posix(), "--wet-run")
    assert caplog.messages[:-1] == snapshot
    assert caplog.messages[-1].startswith(
        "Management command itou.metabase.management.commands.upload_data_to_pilotage succeeded in"
    )
