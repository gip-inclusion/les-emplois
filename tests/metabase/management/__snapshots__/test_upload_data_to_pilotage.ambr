# serializer version: 1
# name: test_command_call
  list([
    "Files in local's 'test_command_call0': ['fluxIAE_ITOU_.tar.gz']",
    "Checking that 'fluxIAE_ITOU_.tar.gz' match with 'flux-iae/fluxIAE_ITOU_.tar.gz'...",
    "'fluxIAE_ITOU_.tar.gz' doesn't match with 'flux-iae/fluxIAE_ITOU_.tar.gz': datastore_content_length=None local_content_length=11",
    "Uploading 'fluxIAE_ITOU_.tar.gz' to 'flux-iae/fluxIAE_ITOU_.tar.gz'...",
    '> fluxIAE_ITOU_.tar.gz: 11\xa0octets/11\xa0octets transferred (100%).',
    "'fluxIAE_ITOU_.tar.gz' match with 'flux-iae/fluxIAE_ITOU_.tar.gz'!",
  ])
# ---
# name: test_command_check_content_length
  list([
    "Files in local's 'test_command_check_content_len0': ['fluxIAE_ITOU_.tar.gz']",
    "Checking that 'fluxIAE_ITOU_.tar.gz' match with 'flux-iae/fluxIAE_ITOU_.tar.gz'...",
    "'fluxIAE_ITOU_.tar.gz' doesn't match with 'flux-iae/fluxIAE_ITOU_.tar.gz': datastore_content_length=0 local_content_length=11",
    "Uploading 'fluxIAE_ITOU_.tar.gz' to 'flux-iae/fluxIAE_ITOU_.tar.gz'...",
    '> fluxIAE_ITOU_.tar.gz: 11\xa0octets/11\xa0octets transferred (100%).',
    "'fluxIAE_ITOU_.tar.gz' match with 'flux-iae/fluxIAE_ITOU_.tar.gz'!",
  ])
# ---
# name: test_command_retries
  list([
    "Files in local's 'test_command_retries0': ['fluxIAE_ITOU_.tar.gz']",
    "Checking that 'fluxIAE_ITOU_.tar.gz' match with 'flux-iae/fluxIAE_ITOU_.tar.gz'...",
    "'fluxIAE_ITOU_.tar.gz' doesn't match with 'flux-iae/fluxIAE_ITOU_.tar.gz': datastore_content_length=None local_content_length=0",
    "Uploading 'fluxIAE_ITOU_.tar.gz' to 'flux-iae/fluxIAE_ITOU_.tar.gz'...",
    "'fluxIAE_ITOU_.tar.gz' doesn't match with 'flux-iae/fluxIAE_ITOU_.tar.gz': datastore_content_length=None local_content_length=0",
    "Uploading 'fluxIAE_ITOU_.tar.gz' to 'flux-iae/fluxIAE_ITOU_.tar.gz'...",
    "'fluxIAE_ITOU_.tar.gz' doesn't match with 'flux-iae/fluxIAE_ITOU_.tar.gz': datastore_content_length=None local_content_length=0",
    "Uploading 'fluxIAE_ITOU_.tar.gz' to 'flux-iae/fluxIAE_ITOU_.tar.gz'...",
    "'fluxIAE_ITOU_.tar.gz' still doesn't match with 'flux-iae/fluxIAE_ITOU_.tar.gz' after 3 tries.",
  ])
# ---
